fulltext/__init__.py

```diff
@@ -1,339 +1,166 @@
-    add_handler() and add_commands().
+    Handle a file-like object.
+
+    Called by `get()` when provided a file-like. This function will prefer the
+    backend's `_get_file()` if one is provided. Otherwise, it will write the
+    data to a temporary file and call `_get_path()`.
     """
-    add_type(mime, ext)
-    add_handler(mime, handler, encoding=encoding)
-    if commands:
-        add_commands(mime, commands, encoding=encoding)
+    if callable(getattr(backend, '_get_file', None)):
+        return backend._get_file(f, **kwargs)
 
+    elif callable(getattr(backend, '_get_path', None)):
+        LOGGER.warning("Using disk, backend does not provde `_get_file()`")
 
-def get_type(filename):
-    """
-    Gets the mimetype and encoding using the mimetypes module. Defined as a standalone
-    function for future expansion.
-    """
-    return mimetypes.guess_type(filename)
+        ext = ''
+        if 'ext' in kwargs:
+            ext = '.' + kwargs['ext']
 
+        with tempfile.NamedTemporaryFile(dir=FULLTEXT_TEMP, suffix=ext) as t:
+            shutil.copyfileobj(f, t)
+            t.flush()
+            return backend._get_path(t.name, **kwargs)
 
-# A placeholder for a kwarg default value.
-class NoDefault(object):
-    pass
+    else:
+        BackendError('Backend must provide `_get_path()` or `_get_file()`')
 
 
-def get(f, default=NoDefault, filename=None, type=None):
+def get(path_or_file, default=SENTINAL, mime=None, name=None, **kwargs):
     """
-    Gets text from a given file. The first parameter can be a path or a file-like object that
-    has a read method. Default is a way to supress errors and just return the default text.
-    Filename can help figure out the type of file being used if you passed in a file-like
-    object. Type can be used to override the type guessing. Type should be a tuple: (mimetype, encoding)
-    as returned by the mimetypes library.
+    Get document full text.
 
-    For any recognizable type, a command will be execute to extract the text. Then that text
-    will be post processed to remove redundant whitespace.
+    Accepts a path or file-like object. If given, `default` is returned
+    instead of an error. `mime` and `name` should be passed if the information
+    is available to caller, otherwise a best guess is made.
+    """
+    if not name:
+        name = getattr(path_or_file, 'name', None)
 
-    For zip files, the first member is extracted and post processed.
+    if not name and isinstance(path_or_file, string_types):
+        name = basename(path_or_file)
 
-    Any file whose type cannot be determined will simply be read then post processed.
-    """
-    if not isinstance(f, basestring) and filename is None:
-        # Try to help figure out the file type.
-        filename = getattr(f, 'name', '')
+    try:
+        backend_name = kwargs.pop('backend')
+    except KeyError:
+        if name:
+            ext = splitext(name)[1].lstrip('.')
```

Comment:
Would a file without an extension through an exception here?

Reply:
No, the extension is an empty string in that case, and strip is happy to strip nothing from it :-).

```python
>>> from os.path import splitext
>>> splitext('foobar')
('foobar', '')
>>> splitext('foobar')[1].lstrip('.')
''
>>> splitext('foobar.txt')
('foobar', '.txt')
>>> splitext('foobar.txt')[1].lstrip('.')
'txt'
```